grails-app/services/org/pih/warehouse/core/DocumentTemplateService.groovy

```diff
@@ -192,4 +195,76 @@ class DocumentTemplateService {
         context.put("orderAdjustments", orderAdjustments);
         return context
     }
+
+    // TODO: make it generic for requistions and orders
+    def renderRequisitionDocumentTemplate(Document documentTemplate, Requisition requisitionInstance, ConverterTypeTo targetDocumentType, OutputStream outputStream) {
+        try {
+            InputStream inputStream = new ByteArrayInputStream(documentTemplate.fileContents)
+            IXDocReport report = XDocReportRegistry.getRegistry().loadReport(inputStream, TemplateEngineKind.Freemarker)
+
+            IContext context = createRequisitionContext(report, requisitionInstance)
+
+            ConverterTypeVia sourceDocumentType = (report.kind == "DOCX") ? ConverterTypeVia.DOCX4J :
+                (report.kind == "ODT") ? ConverterTypeVia.ODFDOM : ConverterTypeVia.XWPF
+
+            // Convert to the target type
+            if (targetDocumentType && sourceDocumentType) {
+                Options options = Options.getTo(targetDocumentType).via(sourceDocumentType)
+                report.convert(context, options, outputStream)
+            }
+            // Process document in place
+            else {
+                report.process(context, outputStream)
+            }
+            log.info "Report dump: " + report.dump()
+        } finally {
+            PdfUtil.restoreBaseFonts()
+        }
+    }
+
+    private createRequisitionContext(IXDocReport report, Requisition requisitionInstance) {
+        def requisitionItems = requisitionInstance?.requisitionItems?.collect { RequisitionItem requisitionItem ->
+            def demand
+            def quantityOnHand
+            if (requisitionInstance?.destination?.isWard()) {
```

Comment:
This logic makes me uneasy since there's no way to tell if demand is for a single location or it's aggregated. I think if there's we should only care about demand for the origin/destination pair and demand should be 0 if a demand record does not exist for that pair. 

Can you ask Kelsey about this?

Reply:
One thing worth mentioning - if it was meant to be working for locations without `MANAGE_INVENTORY` and with `SUBMIT_REQUEST`, it won't, because `isWard()` checks only if `locationTypeCode == LocationTypeCode.WARD`

Reply:
@jmiranda Kelsey wanted to have it working like on add items page. Where it is working like that - if a request is from "ward" then pull demand from the origin to the requestor, and if a request is not from the "ward", then pull demand outgoing from the requestor to all other locations.

@kchelstowski I specifically asked if this is fine on Friday's standup because we are using exactly this in the `stockMovementService.getAddPageItem` and it was fine.